### PR TITLE
Issue 6992: Exercise KVTables, Large event and Streamtags related API.

### LINKED ISSUE #6992: Exercise Apis for KV tables, large event and stream tags
**Is your feature request related to a problem? Please describe.**
This will increase capability of Compatibility Checker class by adding some more checks related to the KV tables, large event, and stream tags.

**Additional context**
For more information please have a look to this JIRA ticket. 
https://jira.cec.lab.emc.com/browse/KARMAN-3035

### PR DESCRIPTION
**Pull request by a6dulaleem**
**Change log description**  
Extends the capability of CompatibilityChecker class it generally does this :-

1. Adding method to write and read large events (16MB) and validating them.
2. Exercising Key Value tables related API and validating them.
3. Its trying to create some streams and putting stream tags and validating them.

**Purpose of the change**  
 Closes #6992 

**What the code does**  
Here in this PR  trying to extend the capability of compatibility checker class by exercising the KVTables, Large event and Stream tags related API.
created specific methods to exercise related api to each topics. 

**How to verify it**  
1- start pravega standalone by ./gradlew startStandalone
2- test the compatibility by running ./gradlew compatibilityCheck -DcontrollerUri=tcp://localhost:9090

### CODE REVIEW

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 1]
```diff
@@ -324,6 +414,153 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+            Preconditions.checkArgument(s.length() <= maxLength);
+            s = Strings.padStart(s, this.maxLength, ' ');
+            return ByteBuffer.wrap(s.getBytes(StandardCharsets.US_ASCII));
+        }
+
+        String deserialize(ByteBuffer b) {
+            Preconditions.checkArgument(b.remaining() <= maxLength);
+            String s = StandardCharsets.US_ASCII.decode(b).toString();
+            s = s.trim();
+            return s;
```
Comment (anju-c-das) [inline test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java:559]: s can be removes as it is not used elsewhere. 
return statement can be changed to StandardCharsets.US_ASCII.decode(b).toString().trim();

**Review by Bhupender-Y — APPROVED** [review]
LGTM

**Review by RaulGracia — CHANGES_REQUESTED** [review]

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 2]
```diff
@@ -324,6 +414,151 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scopeName, controllerURI);
+        readerGroupManager.createReaderGroup(readerGroupId, readerGroupConfig);
+        @Cleanup
+        EventStreamReader<ByteBuffer> reader =  clientFactory.createReader("reader-1", readerGroupId, new ByteBufferSerializer(), ReaderConfig.builder().build());
+
+        int readCount = 0;
+        EventRead<ByteBuffer> event = null;
+        do {
+            event = reader.readNextEvent(10_000);
```
Comment (ShwethaSNayak) [inline test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java:535]: Is it 10000?

**Comment by ShwethaSNayak** [comment]
If I am not wrong **purpose of change** is #6992

**Comment by ShwethaSNayak** [comment]
Also change Description to have large event size as **8MB instead of 4 MB** as pointed out by Raul.

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 3]
```diff
@@ -324,6 +414,151 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+            assertTrue(testKVTables.exists(toKey(userName + i, USERNAME_SERIALIZER)).join());
+            assertEquals(SERIALIZER.deserialize(testKVTables.get(toKey(userName + i, USERNAME_SERIALIZER)).join().getValue()), testData + i);
+        }
+        // Listing all the entries of kv tables
+        val count = new AtomicInteger(0);
+        testKVTables.iterator()
+                .maxIterationSize(20)
+                .all()
+                .keys()
+                .forEachRemaining(ii -> {
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java]: yeah gave a meaningfulname.

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 4]
```diff
@@ -324,6 +414,151 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+                    for (val user : ii.getItems()) {
+                        val key = USERNAME_SERIALIZER.deserialize(user.getPrimaryKey());
+                        System.out.println(String.format("\t%s", key));
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java]: ok

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 5]
```diff
@@ -324,6 +414,151 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+    private void checkWriteAndReadLargeEvent() {
+        String scopeName = "largeevent-test-scope";
+        String streamName = "largeevent-test-stream";
+        createScopeAndStream(scopeName, streamName);
+        @Cleanup
+        EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(scopeName, ClientConfig.builder().controllerURI(controllerURI).build());
+        @Cleanup
+        EventStreamWriter<ByteBuffer> writer = clientFactory.createEventWriter(streamName, new ByteBufferSerializer(), EventWriterConfig.builder().enableLargeEvents(true).build());
+
+        byte[] payload = new byte[Serializer.MAX_EVENT_SIZE * 4];
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java]: sure

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java [hunk 6]
```diff
@@ -324,6 +414,151 @@ private void checkTransactionReadAndWrite() throws TxnFailedException {
...
+    private void checkWriteAndReadLargeEvent() {
+        String scopeName = "largeevent-test-scope";
+        String streamName = "largeevent-test-stream";
+        createScopeAndStream(scopeName, streamName);
+        @Cleanup
+        EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(scopeName, ClientConfig.builder().controllerURI(controllerURI).build());
+        @Cleanup
+        EventStreamWriter<ByteBuffer> writer = clientFactory.createEventWriter(streamName, new ByteBufferSerializer(), EventWriterConfig.builder().enableLargeEvents(true).build());
+
+        byte[] payload = new byte[Serializer.MAX_EVENT_SIZE * 8];
```
Comment (RaulGracia) [inline test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java:516]: If you look at the actual value of `Serializer.MAX_EVENT_SIZE`:
https://github.com/pravega/pravega/blob/40067eed0438504be9c1ba86c7127ac762c26b29/client/src/main/java/io/pravega/client/stream/Serializer.java#L33

It seems to be 8MB already, so I think that just using `Serializer.MAX_EVENT_SIZE * 2` is enough to exercise the large event workflow.

**Review by RaulGracia — COMMENTED** [review]
One minor comment.

File: test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java (hunk 6 above)
Comment (a6dulaleem) [reply test/integration/src/main/java/io/pravega/test/integration/compatibility/CompatibilityChecker.java]: ok, yeah

**Review by RaulGracia — APPROVED** [review]

**Pull request merged**